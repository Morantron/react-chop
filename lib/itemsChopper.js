'use strict';

exports.__esModule = true;
exports.getItemsRangeToRender = getItemsRangeToRender;
exports.getNextScrollState = getNextScrollState;
exports.getFinalBufferingState = getFinalBufferingState;
exports.getNextBufferingState = getNextBufferingState;
function getItemsRangeToRender(_ref) {
  var itemCount = _ref.itemCount,
      windowCount = _ref.windowCount,
      offset = _ref.offset,
      overscan = _ref.overscan;

  console.assert(offset >= 0, 'Offset must be positive');
  console.assert(!offset || offset < itemCount, 'Offset must be less than itemCount');

  var previousOverscanCount = Math.min(offset, overscan);
  var nextOverscanCount = Math.min(itemCount - (overscan + offset), overscan);

  var itemsCountToRender = Math.min(previousOverscanCount + windowCount + nextOverscanCount, itemCount);
  var renderOffset = Math.max(offset - overscan, 0);

  return Array.from({ length: itemsCountToRender }, function (_, i) {
    return renderOffset + i;
  }).filter(function (i) {
    return i < itemCount;
  });
}

function getNextScrollState(_ref2) {
  var itemCount = _ref2.itemCount,
      windowCount = _ref2.windowCount,
      estimatedItemSize = _ref2.estimatedItemSize,
      overscan = _ref2.overscan,
      currentScrollPosition = _ref2.currentScrollPosition;

  var startIndex = Math.floor(currentScrollPosition / estimatedItemSize);
  var maxIndex = Math.max(0, itemCount - windowCount - overscan);
  var offset = Math.max(0, Math.min(startIndex, maxIndex));
  var burgerCount = Math.max(0, offset - overscan);

  return {
    offset: offset,
    burgerCount: burgerCount
  };
}

function getFinalBufferingState(_ref3) {
  var renderedItemsTotalSize = _ref3.renderedItemsTotalSize,
      renderedItemsCount = _ref3.renderedItemsCount,
      containerSize = _ref3.containerSize;

  var estimatedItemSize = renderedItemsTotalSize / renderedItemsCount;
  var newWindowCount = Math.ceil(containerSize / estimatedItemSize);

  return {
    isBuffering: false,
    windowCount: newWindowCount,
    estimatedItemSize: estimatedItemSize
  };
}

function getNextBufferingState(_ref4) {
  var renderedItemsTotalSize = _ref4.renderedItemsTotalSize,
      renderedItemsCount = _ref4.renderedItemsCount,
      containerSize = _ref4.containerSize,
      itemsCount = _ref4.itemsCount,
      shrink = _ref4.shrink,
      itemCount = _ref4.itemCount;

  var itemSize = renderedItemsTotalSize / renderedItemsCount;
  var nextBump = shrink ? itemCount : Math.round(containerSize / itemSize);

  return function (prevState) {
    return {
      isBuffering: true,
      windowCount: prevState.windowCount + nextBump
    };
  };
}