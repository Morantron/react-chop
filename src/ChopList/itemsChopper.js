// TODO: Simplify contract providing just offset and count. Normalize values
// is not a responsibility of this function
export function getItemsRangeToRender({ itemCount, windowCount, offset, overscan }) {
  // Offset range goes from 0 to itemCount - 1
  const normalizedOffset = Math.max(0, Math.min(offset, itemCount - 1));
  const previousOverscanCount = Math.min(normalizedOffset, overscan);
  const nextOverscanCount = Math.min(itemCount - (overscan + normalizedOffset), overscan);

  const itemsCountToRender = Math.min(previousOverscanCount + windowCount + nextOverscanCount, itemCount);
  const renderOffset = Math.max(normalizedOffset - overscan, 0);

  return Array
    .from({length: itemsCountToRender}, (_, i) => renderOffset + i)
    .filter((i) => i < itemCount);
}
