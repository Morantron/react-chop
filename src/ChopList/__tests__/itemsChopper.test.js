import { getItemsRangeToRender } from '../itemsChopper';

test('base', () => {
  const actual = getItemsRangeToRender({
    itemCount: 30,
    windowCount: 5,
    offset: 0,
    overscan: 5,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('base 2', () => {
  const actual = getItemsRangeToRender({
    itemCount: 30,
    windowCount: 5,
    offset: 15,
    overscan: 5,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('base 3', () => {
  const actual = getItemsRangeToRender({
    itemCount: 5,
    windowCount: 10,
    offset: 0,
    overscan: 5,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('no items', () => {
  const actual = getItemsRangeToRender({
    itemCount: 0,
    windowCount: 10,
    offset: 0,
    overscan: 5,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('no overscan', () => {
  const actual = getItemsRangeToRender({
    itemCount: 10,
    windowCount: 5,
    offset: 2,
    overscan: 0,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('windowCount of 1', () => {
  const actual = getItemsRangeToRender({
    itemCount: 10,
    windowCount: 1,
    offset: 5,
    overscan: 0,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('offset greather than itemCount', () => {
  const actual = getItemsRangeToRender({
    itemCount: 10,
    windowCount: 5,
    offset: 11,
    overscan: 0,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('offset less than 0', () => {
  const actual = getItemsRangeToRender({
    itemCount: 10,
    windowCount: 5,
    offset: -2,
    overscan: 0,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('windowCount equal to offset', () => {
  const actual = getItemsRangeToRender({
    itemCount: 1,
    windowCount: 1,
    offset: 1,
    overscan: 0,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});

test('overscan greather than itemCount', () => {
  const actual = getItemsRangeToRender({
    itemCount: 5,
    windowCount: 2,
    offset: 0,
    overscan: 10,
  });

  expect(JSON.stringify(actual)).toMatchSnapshot();
});
