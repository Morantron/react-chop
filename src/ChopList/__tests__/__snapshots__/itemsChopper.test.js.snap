// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base 1`] = `"[0,1,2,3,4,5,6,7,8,9]"`;

exports[`base 2 1`] = `"[10,11,12,13,14,15,16,17,18,19,20,21,22,23,24]"`;

exports[`base 3 1`] = `"[0,1,2,3,4]"`;

exports[`base 4 1`] = `"[]"`;

exports[`no items 1`] = `"[]"`;

exports[`no overscan 1`] = `"[2,3,4,5,6]"`;

exports[`offset greather than itemCount 1`] = `"[9]"`;

exports[`offset less than 0 1`] = `"[0,1,2,3,4]"`;

exports[`overscan greather than itemCount 1`] = `"[]"`;

exports[`windowCount equal to offset 1`] = `"[0]"`;

exports[`windowCount of 1 1`] = `"[5]"`;
